pkg/objstore/inmem/inmem.go

```diff
@@ -34,18 +39,40 @@ func (b *Bucket) Objects() map[string][]byte {
 func (b *Bucket) Iter(_ context.Context, dir string, f func(string) error) error {
 	unique := map[string]struct{}{}
 
+	var dirPartsCount int
+	dirParts := strings.SplitAfter(dir, objstore.DirDelim)
+	for _, p := range dirParts {
+		if p == "" {
+			continue
+		}
+		dirPartsCount++
+	}
 	for filename := range b.objects {
-		if !strings.HasPrefix(filename, dir) {
+		if !strings.HasPrefix(filename, dir) || dir == filename {
 			continue
 		}
-		parts := strings.SplitAfter(filename, "/")
-		unique[parts[0]] = struct{}{}
+
+		parts := strings.SplitAfter(filename, objstore.DirDelim)
+		unique[strings.Join(parts[:dirPartsCount+1], "")] = struct{}{}
 	}
+
 	var keys []string
 	for n := range unique {
 		keys = append(keys, n)
 	}
-	sort.Strings(keys)
+	sort.Slice(keys, func(i, j int) bool {
```

Comment:
Why do we need this ordering here? Also, why can't we just use normal sort?

Reply:
To match ordering of GCS. I decided to took  that as reference. We can choose our own ordering, but we would end up changing order for GCS code instead. The thing is that Thanos works well with GCS so did not want to change ordering (: